### ISSUE TITLE
Why yolov8 automatically start training when invoking eval() and how to turn it off ? 

### ISSUE THREAD

**Issue by Arktische** [ultralytics/ultralytics#9112]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Thanks for your awesome work! I have other models to integrate with yolov8. But when I tried to export the merged model to onnx, it automatically started tarining. I wonder if there is a way to turn this behavior off.

**Comment by glenn-jocher** [comment]
@Arktische hello there! 🌟 Thank you for your kind words and for bringing up your question. In YOLOv8, training behavior kicking in unexpectedly is indeed not typical when performing tasks like exporting models.

The automatic start of training you're observing when invoking `eval()` or related functions might be due to a misunderstanding or a misconfiguration in your code. When exporting models, especially when combining them or performing evaluations, there should be no automatic trigger for the training process.

Could you please double-check your method calls? For exporting a model to ONNX, you should be using something similar to:

```python
from ultralytics import YOLO

# Assuming 'model' is your YOLOv8 model
model = YOLO('path/to/your/model.pt')
model.export(format='onnx', path='path/to/save/onnx/model')
```

If the above doesn't address your issue or if you're doing something similar without explicitly invoking any training functions, it's possible there might be a deeper issue at play or a bug. In such a case, could you please share a snippet of the code where this behavior occurs? This will help in diagnosing the issue more accurately.

Let us know how it goes or if you have any more questions!

**Comment by Arktische** [comment]
@glenn-jocher 
Thanks for your kind reply! Here's minimal reproduction code for this behavior.

```python
from ultralytics import YOLO


YOLOv8 = YOLO("yolov8n.pt")
model = YOLOv8.eval()
```

And I got this output and seemed to start training. Please let me know if I do anyting wrong.

```bash
Ultralytics YOLOv8.1.29 🚀 Python-3.8.18 torch-1.11.0 CUDA:0 (NVIDIA GeForce GTX 1050, 4096MiB)
WARNING ⚠️ Upgrade to torch>=2.0.0 for deterministic training.
engine\trainer: task=detect, mode=train, model=yolov8n.pt, data=coco8.yaml, epochs=100, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=None, name=train7, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=runs\detect\train7

                   from  n    params  module                                       arguments
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]
 16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]
 21                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]
 22        [15, 18, 21]  1    897664  ultralytics.nn.modules.head.Detect           [80, [64, 128, 256]]
Model summary: 225 layers, 3157200 parameters, 3157184 gradients, 8.9 GFLOPs
...
```

**Comment by glenn-jocher** [comment]
Hey @Arktische, glad you reached out with your code! 🌈 It looks like the behavior you're encountering is not related to starting training but instead is how `eval()` is usually expected to function for model evaluation phases, not training.

However, I noticed in your minimal reproduction code and subsequent output that the confusion might stem from the detailed model summary printout that occurs during the model loading phase, which is standard and not indicative of training commencement. This detailed printout includes general warnings and model parameter listings, which are typical when loading the model, especially in verbose mode.

For exporting models, as you intended, this shouldn't pose a problem. If your goal is specifically to export the model to ONNX (or any other supported format) without entering the training phase, ensure that the `export()` function follows the initial model loading, like so:

```python
from ultralytics import YOLO

# Ensure your model is in evaluation phase
YOLOv8 = YOLO("yolov8n.pt").eval()

# Example of exporting the model to ONNX
YOLOv8.export(format='onnx', path='path/to/save/onnx/model')
```

The `eval()` here sets the model to evaluation mode, ideally reducing or eliminating training-specific behaviors such as Dropout or BatchNorm updates. The subsequent `export()` command is what you'd use to convert your model into the desired format, bypassing any training procedures.

If there's anything else we can help you with, or if you have further concerns, feel free to let us know! 🚀

**Comment by mertkiray** [comment]
So, I also observe this problem. Basically I have a pytorch lightning module and inside this pytorch lightning module I have a yolo model. What I want to do is freeze this model and use it only for inference and the resulting predictions as a data for the next model. The problem is if I register the yolo as a child module of my lightning module, when the trainer calls .eval() or .train() methods, the yolo trainer is initialized and it tries to train yolo. The hacky way to get rid of this problem is  to wrap the yolo model inside a list so the lightning does not register it as a child module.  @Arktische  this may also solve your problem.  

Other than this hacky way, what do you suggest? @glenn-jocher

**Comment by glenn-jocher** [comment]
@mertkiray hi there! 🌟 It sounds like you're working on something really interesting by integrating YOLOv8 within a PyTorch Lightning module for inference as part of a larger pipeline. You're correct that using the model in a list is a quick workaround to prevent PyTorch Lightning from auto-registering it as a submodule, which in turn initiates unintentional training behavior through `.eval()` or `.train()`.

A more structured way to include the YOLO model for inference without it being affected by the parent module's training or evaluation state changes could involve overriding the `forward` method of your Lightning module. You could manually manage the YOLO model's state there. Here’s a quick example:

```python
class MyLightningModule(pl.LightningModule):
    def __init__(self, yolov8_model):
        super().__init__()
        self.yolov8_model = yolov8_model
    
    def forward(self, x):
        # Ensure YOLO model is always in eval mode for inference
        self.yolov8_model.eval()
        with torch.no_grad():
            predictions = self.yolov8_model(x)
        # Continue with your pipeline using predictions
        return predictions
```

By doing so, you explicitly keep the YOLO model in evaluation mode (`eval`) right before making predictions, ensuring it doesn't switch to `train` mode inadvertently due to external calls.

Hope that helps! If you have more questions or run into bumps, feel free to reach out. Happy coding! 🚀

**Comment by mertkiray** [comment]
Hey @glenn-jocher, thanks for the response!

The issue I'm encountering is that my Lightning module's forward function isn't being accessed. Oddly, when I keep the self.estimator assignment outside of a list, the training of YOLO initiates. I attempted setting it to evaluation mode as you suggested, but unfortunately, it didn't resolve the problem. It seems like there might be some hidden behavior within Lightning causing this.

```python
class Detector(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.yolo = ultralytics.YOLO('yolov8m.pt')

    def forward(self, images):
        yolo_results = self.yolo.predict(...)

        # Further processing...
```
```python
class Estimator(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.detector = Detector()

    def get_detections(self, image):
        return self.detector(image)

```

```python
class LitModule(LightningModule):
    def __init__(self):
        super().__init__()
        self.estimator = Estimator()

    def forward(self, video):
        with torch.no_grad():
            poses = self.estimator.get_detections(image)
        pred = other_pytorch_model(poses)
        return pred
```

For instance, creating the estimator as `self.estimator = Estimator()` triggers the YOLO training, whereas setting it as `self.estimator = [Estimator()] `works correctly. However, I can't set it to yolo.eval() because it never enters the forward or get_detections() function.

**Comment by glenn-jocher** [comment]
Hey there! 🌈 It seems like the issue might stem from how PyTorch Lightning interacts with the underlying models directly instantiated in the LightningModule. Here’s a quick idea that might help address this without resorting to the list hack:

Since direct instantiation within `__init__` may inadvertently trigger undesired behavior due to how Lightning manages the module lifecycle, you could try wrapping your model instantiation in a method decorated with ` @mertkiray.no_grad()`, ensuring that it’s not influenced by any training processes. Additionally, this method could set the YOLO model to eval mode explicitly.

Here’s a slight modification to your code that you might find helpful:

```python
class LitModule(LightningModule):
    def __init__(self):
        super().__init__()
        self.estimator = self.create_estimator()

    @torch.no_grad()
    def create_estimator(self):
        model = Estimator()
        # Assuming yolo is a model in your Estimator and it's accessible
        model.detector.yolo.eval()
        return model

    def forward(self, video):
        poses = self.estimator.get_detections(video)
        pred = other_pytorch_model(poses)
        return pred
```

This adjustment ensures that your YOLO model is explicitly set to evaluation mode right after its creation, maintaining its state regardless of the parent module's state. The `@torch.no_grad()` decorator prevents unnecessary grad tracking from the start. 

Hope this helps address the challenge! Let me know if you have any more queries or need further assistance. Happy coding! 🚀

**Comment by royvelich** [comment]
I also have this issue. Any recommended way to solve it?

**Comment by glenn-jocher** [comment]
Please ensure you're using the latest version of the Ultralytics package. If the issue persists, try explicitly setting the YOLO model to evaluation mode within your module. If you need further assistance, feel free to provide more details about your setup.

**Comment by mahenning** [comment]
The problem here is that `.eval()` calls [this torch line](https://github.com/pytorch/pytorch/blob/de4c2a3b4e89d96334dc678d1c3f2ae51a6630a0/torch/nn/modules/module.py#L2865): 
```python
return self.train(False)
```
which normally would not be a problem. Torch's `train(True/False)` method handles flipping the training mode on/off (see [here the torch code](https://github.com/pytorch/pytorch/blob/de4c2a3b4e89d96334dc678d1c3f2ae51a6630a0/torch/nn/modules/module.py#L2827)),  
But YOLO overwrites the `train()` method, which now starts the training process! It is in my eyes just bad design to inherit from torch, and overwrite one of its core methods with a completely different behaviour. Just call your training call `start_train` or whatever, not overwriting torchs `train()` method.

Funny enoth, your "solution" with 
```python
@torch.no_grad()
    def create_estimator(self):
        model = Estimator()
        # Assuming yolo is a model in your Estimator and it's accessible
        model.detector.yolo.eval()
        return model
```
still starts the training call from YOLO with `model.eval()`. It just leads to very unexpected behaviour.

**Comment by Y-T-G** [comment]
@mahenning The YOLO class shouldn't be treated like a `torch` model. There's no need to call `.eval()` or `.train()` on it. The correct way to use a class' method is by referring to its documentation and not by assuming that the methods of a different class applies to it.

**Comment by Y-T-G** [comment]
Also the model that ought to be used for export as in the question is `model.model` which has the intended behavior of a PyTorch model.

```python
>>> model = YOLO("yolo11n.pt")
>>> model.model?
Signature:      model.model(*args, **kwargs)
Type:           DetectionModel
String form:
DetectionModel(
           (model): Sequential(
           (0): Conv(
           (conv): Conv2d(3, 16, kernel_size=(3, <...>
           (conv): Conv2d(16, 1, kernel_size=(1, 1), stride=(1, 1), bias=False)
           )
           )
           )
           )
File:           /ultralytics/ultralytics/nn/tasks.py
Docstring:      YOLOv8 detection model.
Init docstring: Initialize the YOLOv8 detection model with the given config and parameters.
```

**Comment by mahenning** [comment]
> The YOLO class shouldn't be treated like a torch model

Then don't inherit from `nn.Module`, or overwrite `.eval()` with `raise NotImplementedError`, or at least a warning. But to inherit from `nn.Module`, have an attribute `.model`, which IS a pytorch model, and starting a training process (and downloading COCO data) when calling `model.eval()` is confusing to the user, to say the least.

I just want to say, I have no (or at least not a big) problem with you using `model.train()` for your training call, but I have a problem with unexpected behavior of other methods like `eval()`, which are not in your documentation. And in my eyes you are also responsible for the behavior of inherited methods. To put it in exaggerated terms, to inherit from a class, overwriting some of its methods with different behavior, breaking other parts saying "just don't touch that, it's not supposed to be used like that" in a github issue is not good design for me.

**Comment by Y-T-G** [comment]
@mahenning Inheriting from `nn.Module` is useful for several reasons such as pickling the class with `torch` or creating a `state_dict`.

I think the `eval()` method behaving like that is a side-effect of `train()` being overridden since `eval()` calls `self.train(False)`. You can open a PR with your suggested fix such adding a `NotImplementedError`.

**Comment by ddilbazTT** [comment]
I ran into this issue while running YOLOv10. The way I handled it is (roughly):

```
# initialize model i.e.
# model = YOLOv10("yolov10.pt")
model = model.model
image = cv2.imread(load_path)
# image preprocessing:
image_rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
input_size = 640
resized = cv2.resize(image_rgb, (input_size, input_size))
normalized = resized.astype(np.float32) / 255.0
x = torch.from_numpy(normalized).permute(2, 0, 1).unsqueeze(0)
output = model(x)
```

The pre-processing was followed from `ultralytics/nn/tasks`. I followed the comments here, they were really helpful. Thanks to everyone who contributed.
